pkg/ingester/ingester.go

```diff
@@ -920,9 +926,7 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) e
 	i.updateMetricsFromPushStats(userID, group, &stats, req.Source, db, i.metrics.discarded)
 
 	if firstPartialErr != nil {
-		code := http.StatusBadRequest
-		errWithUser := wrapOrAnnotateWithUser(firstPartialErr, userID)
-		return newErrorWithHTTPStatus(errWithUser, code)
+		return wrapOrAnnotateWithUser(firstPartialErr, userID)
```

Comment:
Previously we would always return HTTP Status code 400 for ANY error here, even if it wasn't "safe to wrap". After this change, any error that's "not safe to wrap" will be translated to 500. Is that correct?

Reply:
> All the errors returned here were safe,

> These errors are always safe

I don't think this is true. If it is, let's use type system to make sure. But I think there are possible errors that can be returned by `pushSamplesToAppender` that are not wrapped into our "safe" wrappers.

Reply:
In the current implementation, `updateFirstPartial` ([here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L856-L859)) is responsible for assigning a value to variable `firstPartialErr`, that is then retrurned with the 400 HTTP status code ([here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L922-L926)).

The `updateFirstPartial` is passed as a parameter to `pushSamplesToAppender()`, and is called different times inside its implementation (e.g., [here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L982-L984), [here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L989-L992), etc...) 

Related to this PR, all the calls to `updateFirstPartial` in `pushSamplesToAppend()` return what used to be called a `safeToWrap` error, and what now became an `ingesterError`, i.e., an error that `wrapOrAnnotateWithUserId()` would actually wrap with userID. Moreover, `errorType()` of all these `ingesterError`s is `ingesterErrorType.badData`, i.e., the type of errors that is mapped to 400.

Reply:
> In the current implementation, `updateFirstPartial` ([here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L856-L859)) is responsible for assigning a value to variable `firstPartialErr`, that is then retrurned with the 400 HTTP status code ([here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L922-L926)).
> 
> The `updateFirstPartial` is passed as a parameter to `pushSamplesToAppender()`, and is called different times inside its implementation (e.g., [here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L982-L984), [here](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L989-L992), etc...)

I see, and I think you're right. It would be great if we could codify that, for example by defining `updateFirstPartial` so that it only accepts `ingesterError`s.

Reply:
We could also pass `*log.Sampler` to `updateFirstPartial` to avoid problem with `SampledError` wrapping. 🤔

Reply:
I don't understand which problem with `SampledError` wrapping you are referring to. Anyway, I don't think it is a good idea to pass a `*log.Sampler` to `updateFirstPartial` because there are some errors that don't need to be sampled, for example [this](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L1199-L1201).

Reply:
> I see, and I think you're right. It would be great if we could codify that, for example by defining `updateFirstPartial` so that it only accepts `ingesterError`s.

I can try to do it, but what do we actually want to achieve? If the error returned by `updateFirstPartial()`'s parameter is something we consider a silent error, we go on, but if it is another type of error we want to fail immediately?

Reply:
> but what do we actually want to achieve? If the error returned by updateFirstPartial()'s parameter is something we consider a silent error, we go on, but if it is another type of error we want to fail immediately?

What we want to achieve is to make sure that we don't "leak" any other types of errors via `firstPartialErr` (that can't be safely wrapped). As stated previously, even if that was the case before this PR, we would still end up returning 400, but after this PR that's not the case anymore -- instead call to `wrapOrAnnotateWithUser` with unwrappable error will "destroy" underlying error type and our mapping will not convert it to anything.

While we can verify that only "ingester errors" are returned `firstPartialErr` today by doing a code review, it would be better to actually use the type system to avoid any future "leaks" of non-ingester errors here.

I agree about doing this in separate PR.

Reply:
> I don't understand which problem with SampledError wrapping you are referring to.

I've tried to change `updateFirstPartial` to only accept `ingesterError`, but that wasn't working for errors wrapped by `SampledError`.

> Anyway, I don't think it is a good idea to pass a *log.Sampler to updateFirstPartial because there are some errors that don't need to be sampled, for example [this](https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L1199-L1201).

I don't think that's a problem -- we would pass `nil` in such case.

Reply:
>I don't think that's a problem -- we would pass nil in such case.

I have updated the `updateFirstPartial` signature accordingly. I will do additional changes in a separate PR.